dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1062 @@
+   * Subscript operator, returns global reference to element at given
+   * global index.
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  reference operator[](index_type idx) const {
+    auto coords    = glob_coords(idx);
+    auto local_pos = _pattern.local_index(coords);
+
+    return reference(_globmem.at(local_pos.unit, local_pos.index));
+  }
+
+  dart_gptr_t dart_gptr() const { return operator[](_idx).dart_gptr(); }
+
+  /**
+   * Checks whether the element referenced by this global iterator is in
+   * the calling unit's local memory.
+   */
+  bool is_local() const { return (_myid == lpos().unit); }
+
+  GlobIter<ElementT, PatternT> global() const {
+    auto g_idx = gpos();
+    return GlobIter<ElementT, PatternT>(&_globmem, _pattern, g_idx);
+  }
+
+  ElementT* local() const {
+    auto local_pos = lpos();
+
+    if (_myid != local_pos.unit)
+      return nullptr;
+
+    //
+    return _lbegin + local_pos.index;
+  }
+
+  /**
+   * Position of the iterator in global storage order.
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  index_type pos() const { return gpos(); }
+
+  /**
+   * Position of the iterator in its view's iteration space, disregarding
+   * the view's offset in global index space.
+   *
+   * \see DashViewIteratorConcept
+   */
+  index_type rpos() const { return _idx; }
+
+  /**
+   * Position of the iterator in global index range.
+   * Projects iterator position from its view spec to global index domain.
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  index_type gpos() const { return _pattern.memory_layout().at(glob_coords(_idx)); }
+
+  std::array<index_type, NumDimensions> gcoords() const { return glob_coords(_idx); }
+
+  typename PatternT::local_index_t lpos() const { return _pattern.local_index(glob_coords(_idx)); }
+
+  const ViewSpecT viewspec() const { return _region_view; }
+
+  inline bool is_relative() const noexcept { return true; }
+
+  /**
+   * The instance of \c GlobStaticMem used by this iterator to resolve addresses
+   * in global memory.
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  const GlobMemT& globmem() const { return _globmem; }
+
+  /**
+   * Prefix increment operator.
+   */
+  SelfT& operator++() {
+    ++_idx;
+    return *this;
+  }
+
+  /**
+   * Postfix increment operator.
+   */
+  SelfT operator++(int) {
+    SelfT result = *this;
+    ++_idx;
+    return result;
+  }
+
+  /**
+   * Prefix decrement operator.
+   */
+  SelfT& operator--() {
+    --_idx;
+    return *this;
+  }
+
+  /**
+   * Postfix decrement operator.
+   */
+  SelfT operator--(int) {
+    SelfT result = *this;
+    --_idx;
+    return result;
+  }
+
+  SelfT& operator+=(index_type n) {
+    _idx += n;
+    return *this;
+  }
+
+  SelfT& operator-=(index_type n) {
+    _idx -= n;
+    return *this;
+  }
+
+  SelfT operator+(index_type n) const {
+    SelfT res{*this};
+    res += n;
+
+    return res;
+  }
+
+  index_type operator+(const SelfT& other) const { return _idx + other._idx; }
```

Comment:
Formatting! 

Plus: why does this `operator+` return `index_type`?

Same holds true for the below operator.